File: src/cadforge/chat/modes.py
"""Interaction modes for CadForge REPL.

Defines agent, plan, and ask modes with their tool sets and permissions.
"""

from __future__ import annotations

from enum import Enum
from typing import FrozenSet


class InteractionMode(Enum):
    """REPL interaction modes."""

    AGENT = "agent"
    PLAN = "plan"
    ASK = "ask"

    @property
    def prompt_prefix(self) -> str:
        return f"cadforge:{self.value}> "

    @property
    def display_name(self) -> str:
        return self.value.upper()

    @property
    def color(self) -> str:
        return _MODE_COLORS[self]

    def next(self) -> InteractionMode:
        """Cycle to the next mode: agent -> plan -> ask -> agent."""
        order = [InteractionMode.AGENT, InteractionMode.PLAN, InteractionMode.ASK]
        idx = order.index(self)
        return order[(idx + 1) % len(order)]


_MODE_COLORS: dict[InteractionMode, str] = {
    InteractionMode.AGENT: "green",
    InteractionMode.PLAN: "yellow",
    InteractionMode.ASK: "cyan",
}

# Read-only tools allowed in plan mode
PLAN_MODE_TOOLS: frozenset[str] = frozenset(
    {"ReadFile", "ListFiles", "SearchVault", "Bash", "GetPrinter", "Task"}
)


def get_mode_tools(mode: InteractionMode) -> frozenset[str] | None:
    """Return the set of allowed tool names for a mode.

    Returns None for agent mode (all tools allowed),
    a frozenset for plan mode (read-only tools),
    or an empty frozenset for ask mode (no tools).
    """
    if mode == InteractionMode.AGENT:
        return None
    if mode == InteractionMode.PLAN:
        return PLAN_MODE_TOOLS
    return frozenset()


PROCEED_PHRASES: frozenset[str] = frozenset({
    "proceed", "go ahead", "do it", "execute", "implement",
    "build it", "make it", "let's go", "start", "run it",
})


def has_proceed_intent(user_input: str) -> bool:
    """Return True if the user input signals intent to execute a plan."""
    lower = user_input.lower().strip()
    return any(phrase in lower for phrase in PROCEED_PHRASES)


def get_plan_mode_permissions() -> dict[str, list[str]]:
    """Return permissions dict that denies write tools and allows read tools."""
    return {
        "deny": [
            "WriteFile(*)",
            "ExecuteCadQuery(*)",
            "ExportModel(*)",
            "AnalyzeMesh(*)",
            "ShowPreview(*)",
            "SearchWeb(*)",
        ],
        "allow": [
            "ReadFile(*)",
            "ListFiles(*)",
            "SearchVault(*)",
            "GetPrinter(*)",
            "Task(*)",
            "Bash(*)",
        ],
        "ask": [],
    }

File: packages/cli/src/agent/modes.ts
/**
 * Interaction modes for the REPL.
 *
 * Defines agent, plan, and ask modes with their tool sets and permissions.
 */

import type { PermissionsConfig } from '@cadforge/shared';

export type InteractionMode = 'agent' | 'plan' | 'ask';

export const MODE_COLORS: Record<InteractionMode, string> = {
  agent: 'green',
  plan: 'yellow',
  ask: 'cyan',
};

export function nextMode(mode: InteractionMode): InteractionMode {
  const order: InteractionMode[] = ['agent', 'plan', 'ask'];
  const idx = order.indexOf(mode);
  return order[(idx + 1) % order.length];
}

export function modePromptPrefix(mode: InteractionMode): string {
  return `cadforge:${mode}> `;
}

/** Read-only tools allowed in plan mode */
const PLAN_MODE_TOOLS = new Set([
  'ReadFile', 'ListFiles', 'SearchVault', 'Bash', 'GetPrinter', 'Task',
]);

/**
 * Return the set of allowed tool names for a mode.
 * Returns null for agent mode (all tools allowed).
 */
export function getModeTools(mode: InteractionMode): Set<string> | null {
  if (mode === 'agent') return null;
  if (mode === 'plan') return PLAN_MODE_TOOLS;
  return new Set(); // ask: no tools
}

const PROCEED_PHRASES = new Set([
  'proceed', 'go ahead', 'do it', 'execute', 'implement',
  'build it', 'make it', "let's go", 'start', 'run it',
]);

export function hasProceedIntent(userInput: string): boolean {
  const lower = userInput.toLowerCase().trim();
  for (const phrase of PROCEED_PHRASES) {
    if (lower.includes(phrase)) return true;
  }
  return false;
}

/**
 * Return permissions that deny write tools and allow read tools (for plan mode).
 */
export function getPlanModePermissions(): PermissionsConfig {
  return {
    deny: [
      'WriteFile(*)',
      'ExecuteCadQuery(*)',
      'ExportModel(*)',
      'AnalyzeMesh(*)',
      'ShowPreview(*)',
      'SearchWeb(*)',
    ],
    allow: [
      'ReadFile(*)',
      'ListFiles(*)',
      'SearchVault(*)',
      'GetPrinter(*)',
      'Task(*)',
      'Bash(*)',
    ],
    ask: [],
  };
}
